source/core/slang-common.h

```diff
@@ -76,28 +76,43 @@ namespace Slang
     template <typename T>
     SLANG_FORCE_INLINE T* clone(IClonable* clonable) { return (T*)clonable->clone(T::getTypeGuid()); }
 
+}
 
 // TODO: Shouldn't these be SLANG_ prefixed?
 #ifdef _MSC_VER
 #define UNREACHABLE_RETURN(x)
-#define UNREACHABLE(x)
 #else
 #define UNREACHABLE_RETURN(x) return x;
-#define UNREACHABLE(x) x;
 #endif
 
-}
+//
+// Use `SLANG_ASSUME(myBoolExpression);` to inform the compiler that the condition is true.
+// Do not rely on side effects of the condition being performed.
+//
+#if defined(__cpp_assume)
+#    define SLANG_ASSUME(X) [[assume(X)]]
+#elif SLANG_GCC
+#    define SLANG_ASSUME(X) do{if(!(X)) __builtin_unreachable();} while(0)
+#elif SLANG_CLANG
+#    define SLANG_ASSUME(X) __builtin_assume(X)
+#elif SLANG_VC
+#    define SLANG_ASSUME(X) __assume(X)
```

Comment:
This was exactly the intention. Even if we did have `assert(cond); assume(!cond)` the compiler would be free to optimize away the assert because it wouldn't ever trigger under the assumption of `!cond`. It's not just a hint, it's a promise which if broken means the compiler may do what it pleases.

Reply:
Oh, I see it is used in release only, then that makes sense.